### PR TITLE
Extract MiqStats from manageiq-gems-pending

### PR DESCRIPTION
**Pull request by bdunne**
Add some basic tests

### CODE REVIEW

File: lib/more_core_extensions/core_ext/math/slope.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
+module MoreCoreExtensions
+  module MathSlope
+    module ClassMethods
+      # Finds the x coordinate given y, slope of the line and the y intercept
+      #
+      # x = (y - b) / m
+      # Where m is the slope of the line and b is the y intercept (the location where the line crosses the y axis)
+      def slope_x_intercept(y, slope, y_intercept)
```
Comment (Fryguy) [inline lib/more_core_extensions/core_ext/math/slope.rb]: we may want keyword args here for caller readability, but that can be a follow up PR.

File: spec/core_ext/math/slope_spec.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  it ".slope_y_intercept" do
+    expect(described_class.slope_y_intercept(1, 1, 0)).to eq(1)
+    expect(described_class.slope_y_intercept(2, 1, 0)).to eq(2)
+    expect(described_class.slope_y_intercept(10, 1, 0)).to eq(10)
+    expect(described_class.slope_y_intercept(10, 2, 0)).to eq(20)
+  end
+
+  it ".slope" do
+    expect(described_class.slope([1.0, 2.0], [1.0, 2.0])).to eq([1.0, 0.0, 1.0])
+    expect(described_class.slope([1.0, 1.5], [1.0, 2.0])).to eq([2.0, -1.0, 1.0])
```
Comment (Fryguy) [inline spec/core_ext/math/slope_spec.rb]: I find the inputs so strange.  It's like you decompose the x,y pairs into x's and y's.  I wonder if we should just accept an Array of x/y pairs instead, or if the user wants they could give a Hash, and either way we call .to_a

Comment (Fryguy) [inline spec/core_ext/math/slope_spec.rb]: I assume `m` is slope and `b` is the y-intercept...what is `r`?

Comment (bdunne) [inline spec/core_ext/math/slope_spec.rb]: I agree, the inputs feel strange, I would expect 2 or more x/y pair inputs.

Comment (bdunne) [inline spec/core_ext/math/slope_spec.rb]: I don't understand the output, I was expecting just the slope.  @Fryguy I was hoping you would know 😄

Comment (gtanzillo) [inline spec/core_ext/math/slope_spec.rb]: I was looking at this yesterday and the output array consists of `slope`, `y-intercept` and `correlation coefficient`. We definitely use the `y-intercept` . The correlation coefficient is a measure of the strength of slope calculation based on the inputs. I haven't been able to find a place with it is being used yet. But, I can't say this it's not being used.

Comment (Fryguy) [inline spec/core_ext/math/slope_spec.rb]: ohhhh...this is not a slope, and I think that's what confused me.  This is actually a linear trendline calculation where r is the **r squared** value http://condor.depaul.edu/sjost/it223/documents/correlation.htm

Let's rename this method to `linear_trend`, then we can rename the params appropriately.

**Comment by Fryguy** [comment]
@bdunne @gtanzillo I've updated this PR with better method names and some cleanup for readability.

File: lib/more_core_extensions/core_ext/math/slope.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        x_array, y_array = coordinates.transpose
+        sum_x  = x_array.sum
+        sum_x2 = x_array.map(&:square).sum
+        sum_y  = y_array.sum
+        sum_y2 = y_array.map(&:square).sum
+        sum_xy = coordinates.map { |x, y| x * y }.sum
+
+        n = coordinates.size.to_f
+        slope = (n * sum_xy - sum_x * sum_y) / (n * sum_x2 - sum_x.square)
+        return if slope.nan?
```
Comment (bdunne) [inline lib/more_core_extensions/core_ext/math/slope.rb]: Should this raise?

Comment (gtanzillo) [reply lib/more_core_extensions/core_ext/math/slope.rb]: I don't think so. I think theres code that doesn't validate the values it sends and relies on it not raising

Comment (Fryguy) [reply lib/more_core_extensions/core_ext/math/slope.rb]: I don't think so...it's not an error that the slope doesn't exist, and nil seems like a valid response

For example, the slope of a single point doesn't exist, but it's not an error.  Another example is the slope of 2 points with the same x value (1,1), (1,2); becomes a vertical line, which means the slope doesn't exist.

**Pull request merged**